### DISCUSSION TITLE
How could I exclude null values from an GetArrayValues query?

### DISCUSSION THREAD

**Discussion by ianww** [cables-gl/cables_docs#290] [Answers & Questions] [answered]
I'm using GetArrayValues calls to construct a latitude array and a longitude array.  I then pack those arrays into an ArrayPack2 and use those coordinates to plot values (speed) onto the surface of a sphere.  Both the latitude and longitude arrays contain null values, not necessarily both together at the same index.  The null values appear to get treated as zero, leading to unwanted artefacts around the equator and the zero longitude.

I want to exclude the nulls from being plotted.  This would require iterating through the lat and long arrays together and excluding values from both if either is null at a given index.  I can't find any not-null or exclude op.

I'd appreciate some advice on how I could go about doing this!  

![image](https://user-images.githubusercontent.com/36909839/178140840-c560a9d0-8438-465d-a2ee-60b8539b5ba9.png)

![image](https://user-images.githubusercontent.com/36909839/178139906-ee744b94-ae8f-4e19-b989-ef5ff8ed26a0.png)

**Comment by ianww** [comment]
I'm trying to do it with a user-op using my code below but I'm having trouble with the initialization of the input arrays.  The output from the line 7 console.log call is "_Array 1 is length **undefined** and Array 2 is length **undefined**_".   This is despite a mouse hover showing that the input arrays are populated.

![image](https://user-images.githubusercontent.com/36909839/178205769-b524222d-339c-450d-b27f-26107d4cf3dc.png)

Can someone please help me with why my code is failing to extract the input arrays?

(In case it's important, fyi the user-op is on a trigger-once cable chain because it's processing the results of an API call, which I only want to do once.) 

```
//Create input and output ports and intialize them

const arrayIn_1 = op.inArray("Array in 1");
const arrayIn_2 = op.inArray("Array in 2");
const arrayOut_1 = op.outArray("Array out 1");
const arrayOut_2 = op.outArray("Array out 2");
console.log("Array 1 is length ", arrayIn_1.length, " and Array 2 is length ", arrayIn_2.length);

//Create filter arrays
var arr1_filtered = [];
var arr2_filtered = [];

//Filtering function
function filter()
{
     console.log("Filtering...");
    //Assign the input arrays to local arrays
    var arr1 = arrayIn_1.get();
    var arr2 = arrayIn_2.get();

    //loop through contents of both arrays and check for nulls
    for (var i = 0; i < arr1.length; i++)
    {
        if((arr1[i] == null) || (arr2[i] == null))
        {
            console.log("Null found");
        }
        else //Append the values to the filtered arrays
        {
            arr1_filtered.push(arr1[i]);
            arr2_filtered.push(arr2[i]);
        }
    }
}

//Run the filter
filter();

//Push the filtered arrays to the outputs
arrayOut_1.set(arr1_filtered);
arrayOut_2.set(arr2_filtered);
```

**Reply by steam0r** [reply]
you are not running your filtering "onChange" for the ports, see here: https://dev.cables.gl/docs/5_writing_ops/dev_callbacks/dev_callbacks

this means you are only running it on creation of the op, and hence the array-lenght is undefined. even if you fix this, you should still handle cases where the lenght could be undefined (during init, when connecting/disconnecting) one of the ports in you newer code.

**Reply by ianww** [reply]
Ah, thanks!  I replaced my function header with `arrayIn_1.onChange=arrayIn_2.onChange = function()` (and removed the later filter() call) and it now works.  

I'd wrongly assumed that the array inputs wouldn't change because they were part of a trigger-once call, not realizing that the trigger would be expended creating the op.

I'll also add in a handler for undefined arrays, as you recommend.

Many thanks!

**Reply by ianww** [reply]
Here's the corrected code for my FilterNullsFromArrays user-op, and its use and results.  Artefacts be gone!

```
//Create input and output ports and intialize them

const arrayIn_1 = op.inArray("Array in 1");
const arrayIn_2 = op.inArray("Array in 2");
const arrayOut_1 = op.outArray("Array out 1");
const arrayOut_2 = op.outArray("Array out 2");

//Create filter arrays
var arr1_filtered = [];
var arr2_filtered = [];

//Filtering function
arrayIn_1.onChange=arrayIn_2.onChange = function()
{
     console.log("Filtering...");
    //Assign the input arrays to local arrays
    var arr1 = arrayIn_1.get();
    var arr2 = arrayIn_2.get();

    //If array(s) aren't present set the output arrays to null and return
    if(!arr1 || !arr2)
    {
        arrayOut_1.set(null);
        arrayOut_2.set(null);
        return;
    }

    //Loop through contents of both arrays and check for nulls
    for (var i = 0; i < arr1.length; i++)
    {
        if((arr1[i] == null) || (arr2[i] == null))
        {
            console.log("Null found");
        }
        else //Append the values to the filtered arrays
        {
            arr1_filtered.push(arr1[i]);
            arr2_filtered.push(arr2[i]);
        }
    }
}

//Push the filtered arrays to the outputs
arrayOut_1.set(arr1_filtered);
arrayOut_2.set(arr2_filtered);
```

![image](https://user-images.githubusercontent.com/36909839/178215916-5014b12c-b46c-41f4-9c1c-e6182bf0cab3.png)

![image](https://user-images.githubusercontent.com/36909839/178216260-eb8afb8f-8b50-47e3-9858-e3f57eec265f.png)

**Comment by ianww** [comment, marked as answer]
Although my code worked initially, when I reloaded the patch it failed due to a null array error.  I discovered that the push of the filtered arrays to the outputs needs to be inside the function.  The corrected code is:

```
//Create input and output ports and intialize them

const arrayIn_1 = op.inArray("Array in 1");
const arrayIn_2 = op.inArray("Array in 2");
const arrayOut_1 = op.outArray("Array out 1");
const arrayOut_2 = op.outArray("Array out 2");

//Create filter arrays
const arr1_filtered = [];
const arr2_filtered = [];

//Filtering function
arrayIn_1.onChange=arrayIn_2.onChange=function()
{
    console.log("Filtering...");
    //Assign the input arrays to local arrays
    var arr1 = arrayIn_1.get();
    var arr2 = arrayIn_2.get();

    //If array(s) aren't present set output arrays to null and return
    if(!arr1 || !arr2)
    {
        arrayOut_1.set(null);
        arrayOut_2.set(null);
        console.log("An input array is missing, so both output arrays set to null");
        return;
    }

    //If the arrays have different lengths set output arrays out to null and return
    if(arr1.length != arr2.length)
    {
        arrayOut_1.set(null);
        arrayOut_2.set(null);
        console.log("Arrays have different lengths, so both output arrays set to null");
        return;
    }

    //Loop through contents of both arrays and check for nulls
    for (var i = 0; i < arr1.length; i++)
    {
        if((arr1[i] == null) || (arr2[i] == null))
        {
            console.log("Null found");
        }
        else //Append the values to the filtered arrays
        {
            arr1_filtered.push(arr1[i]);
            arr2_filtered.push(arr2[i]);
        }
    }

//Push the filtered arrays to the outputs
arrayOut_1.set(arr1_filtered);
arrayOut_2.set(arr2_filtered);
}
```
